api/src/main/java/org/openmrs/validator/UserValidator.java

```diff
@@ -35,170 +35,173 @@
+          errors.rejectValue("username", "error.username.email");
+        }
+      } else {
+        boolean isValidUserName = isUserNameValid(user.getUsername());
+        if (!isValidUserName) {
+          errors.rejectValue("username", "error.username.pattern");
+        }
+      }
+
+      if (!StringUtils.isBlank(user.getEmail()) && !isEmailValid(user.getEmail())) {
+        errors.rejectValue("email", "error.email.invalid");
+      } else {
+        ValidateUtil.validateFieldLengths(errors, obj.getClass(), "username", "systemId", "retireReason");
+      }
+    }
+  }
+
+  /**
+   * Convenience method to check the given username against the regular
+   * expression. <br>
+   * <br>
+   * A valid username will have following:
+   * <ul>
+   * <li>Begins with Alphanumeric characters
+   * <li>only followed by more alphanumeric characters (may include . - _)
+   * <li>can be at most 50 characters
+   * <li>minimum 2 chars case-insensitive Examples:
+   * <li>The following username will pass validation: A123_.-XYZ9
+   * </ul>
+   *
+   * @param username the username string to check
+   * @return true if the username is ok <strong>Should</strong> validate username
+   *         with only alpha numerics <strong>Should</strong> validate username
+   *         with alpha dash and underscore <strong>Should</strong> validate
+   *         username with alpha dash underscore and dot <strong>Should</strong>
+   *         validate username with exactly max size name <strong>Should</strong>
+   *         not validate username with less than minimumLength
+   *         <strong>Should</strong> not validate username with invalid character
+   *         <strong>Should</strong> not validate username with more than maximum
+   *         size <strong>Should</strong> validate when username is null
+   *         <strong>Should</strong> validate when username is the empty string
+   *         <strong>Should</strong> not validate when username is whitespace only
+   */
+  public boolean isUserNameValid(String username) {
+    // Initialize reg ex for userName pattern
+    // ^ = start of line
+    // \w = [a-zA-Z_0-9]
+    // \Q = quote everything until \E
+    // $ = end of line
+    // complete meaning = 2-50 characters, the first must be a letter, digit, or _,
+    // and the rest may also be - or .
+    String expression = "^[\\w][\\Q_\\E\\w-\\.]{1,49}$";
+    // empty usernames are allowed
+    if (StringUtils.isEmpty(username)) {
+      return true;
+    }
+
+    try {
+      // Make the comparison case-insensitive.
+      Pattern pattern = Pattern.compile(expression, Pattern.CASE_INSENSITIVE);
+      Matcher matcher = pattern.matcher(username);
+      return matcher.matches();
+    } catch (PatternSyntaxException pex) {
+      log.error("Username Pattern Syntax exception in UserValidator", pex);
+      return false;
+    }
+  }
+
+  /**
+   * Returns true if the given username is a valid e-mail.
+   *
+   * @param username
+   * @return true if valid
```

Comment:
@sherrif10 you seem to have more changes than required by the ticket.

Reply:
That was there before my new changes

Reply:
Is it required by the ticket?

Reply:
Yes sure its required since it was formally there

Reply:
Let me put it differently, does the ticket require making all these changes? https://github.com/openmrs/openmrs-core/pull/3531/files

Reply:
@sacull are you still looking into this. take a look once more

Reply:
@sherrif10 it is best to fist deal with what other reviewer have brought up, before you again ping @sacull 
That way, you do not over burden him.

Reply:
ok thanks @dkayiwa